### PR TITLE
Feature/config loader

### PR DESCRIPTION
**Pull request by Solzc**

### CODE REVIEW

File: rr/config/app_validator.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
+#!/usr/bin/env python3
+#  Copyright (C) 2021 RidgeRun, LLC (http://www.ridgerun.com)
+#  Authors: Daniel Chaves <daniel.chaves@ridgerun.com>
+#           Marisol Zeledon <marisol.zeledon@ridgerun.com>
+
+
+class AppValidatortError(RuntimeError):
```
Comment (emmadrigal) [inline rr/config/app_validator.py:246]: This whole document is missing docstring

File: rr/config/config_loader.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+#  Copyright (C) 2021 RidgeRun, LLC (http://www.ridgerun.com)
+#  Authors: Daniel Chaves <daniel.chaves@ridgerun.com>
+#           Marisol Zeledon <marisol.zeledon@ridgerun.com>
+
+
+from rr.config.app_validator import AppValidator
+from rr.config.yaml_format import YamlFormat
+
+
+class ConfigLoader:
```
Comment (emmadrigal) [inline rr/config/config_loader.py:11]: Missing docstring

File: rr/config/app_validator.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+class AppValidator():
+
+    def validate(self, cfg):
+        self.validate_streams(cfg)
+        self.validate_filters(cfg)
+        self.validate_actions(cfg)
+        self.validate_triggers(cfg)
+
+    def validate_streams(self, cfg):
```
Comment (emmadrigal) [inline rr/config/app_validator.py]: Other than `validate` all the functions in this class don't require `self`, move them to functions

File: rr/config/yaml_format.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env python3
+#  Copyright (C) 2021 RidgeRun, LLC (http://www.ridgerun.com)
+#  Authors: Daniel Chaves <daniel.chaves@ridgerun.com>
+#           Marisol Zeledon <marisol.zeledon@ridgerun.com>
+
+import yaml
+
+
+class YamlFormatError(RuntimeError):
```
Comment (emmadrigal) [inline rr/config/yaml_format.py:9]: Missing docstring

File: tests/config/test_app_validator.py [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+class TestYamlFormat(unittest.TestCase):
+
+    def setUp(self):
+        self.validator = AppValidator()
+
+    def test_streams(self):
+        cfg_obj1 = {'streams': [
+            {'id': 'stream0', 'uri': 'http0', 'triggers': ['person_recording']}]}
+        cfg_obj2 = {'triggers': 'test', 'filters': 0}
+        cfg_obj3 = {'streams': 'test'}
```
Comment (emmadrigal) [inline tests/config/test_app_validator.py]: Try to use more representative names. Something like:
```
cfg_good =
cfg_missing_stream =
cfg_invalid_stream =
```

File: tests/config/test_app_validator.py [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+
```
Comment (emmadrigal) [inline tests/config/test_app_validator.py:18]: Please add a test with `cfg_empty = None`

File: tests/config/test_app_validator.py [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+        with self.assertRaises(AppValidatortError) as e2:
+            self.validator.validate_streams(cfg_obj3)
+
+        self.assertEqual(None, streams_validate)
+        self.assertEqual("streams object not found", str(e1.exception))
+        self.assertEqual("Invalid streams format", str(e2.exception))
+
+    def test_id_errors(self):
+        cfg_obj1 = {'streams': [
```
Comment (emmadrigal) [inline tests/config/test_app_validator.py]: Use more representative names

File: tests/config/test_app_validator.py [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+            self.validator.validate_streams(cfg_obj1)
+
+        with self.assertRaises(AppValidatortError) as e2:
+            self.validator.validate_streams(cfg_obj2)
+
+        self.assertEqual("id property not found in stream", str(e1.exception))
+        self.assertEqual("Invalid id format", str(e2.exception))
+
+    def test_uri_errors(self):
+        cfg_obj1 = {'streams': [
```
Comment (emmadrigal) [inline tests/config/test_app_validator.py]: Use more representative names

File: tests/config/test_app_validator.py [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+            self.validator.validate_streams(cfg_obj1)
+
+        with self.assertRaises(AppValidatortError) as e2:
+            self.validator.validate_streams(cfg_obj2)
+
+        self.assertEqual("uri property not found in stream", str(e1.exception))
+        self.assertEqual("Invalid uri format", str(e2.exception))
+
+    def test_streams_triggers_errors(self):
+        cfg_obj1 = {'streams': [
```
Comment (emmadrigal) [inline tests/config/test_app_validator.py]: Use more representative names

File: rr/config/app_validator.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+            except KeyError as e:
+                raise AppValidatortError(
+                    "triggers property not found in stream") from e
+
+            if (not isinstance(stream_id, str)):
+                raise AppValidatortError("Invalid id format")
+            if (not isinstance(stream_uri, str)):
+                raise AppValidatortError("Invalid uri format")
+            if (not isinstance(stream_triggers, list)):
+                raise AppValidatortError("Invalid triggers format")
```
Comment (emmadrigal) [inline rr/config/app_validator.py]: The message could be more informative.
```
Invalid id format, should be str
...
Invalid triggers format, should be a list
```

File: rr/config/app_validator.py [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+            for trigger in stream_triggers:
+                if (not isinstance(trigger, str)):
+                    raise AppValidatortError(
+                        "Invalid trigger format in streams triggers")
```
Comment (emmadrigal) [inline rr/config/app_validator.py]: ```Invalid trigger format in streams triggers, should be string```

File: rr/config/app_validator.py [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+            except KeyError as e:
+                raise AppValidatortError(
+                    "threshold property not found in filter") from e
+
+            if (not isinstance(filter_name, str)):
+                raise AppValidatortError("Invalid name format in filter")
+            if (not isinstance(filter_labels, list)):
+                raise AppValidatortError("Invalid labels format in filter")
+            if (not isinstance(filter_thresh, float)):
+                raise AppValidatortError("Invalid threshold format")
```
Comment (emmadrigal) [inline rr/config/app_validator.py]: Add more info to error message `should be ...`

File: rr/config/app_validator.py [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+        if (not isinstance(action_name, str)):
+            raise AppValidatortError("Invalid name format in action")
+        if (not isinstance(action_type, str)):
+            raise AppValidatortError("Invalid type format in action")
+        if (action_type == 'recording'):
+            if (not isinstance(recording_lenght, int)):
+                raise AppValidatortError(
+                    "Invalid lenght format in action of type recording")
+        if (not isinstance(action_location, str)):
+            raise AppValidatortError("Invalid location format in action")
```
Comment (emmadrigal) [inline rr/config/app_validator.py]: Add more info to error message about correct type

File: tests/config/test_app_validator.py [hunk 14]
```diff
@@ -0,0 +1,415 @@
...
+                            {'name': 'person_logging',
+                             'action': 'log_event',
+                             'filters': ['person_filter']},
+                            {'name': 'animal_logging',
+                             'action': 'log_event',
+                             'filters': ['animal_filter']}]}
+
+        validate = self.validator.validate(cfg)
+
+        self.assertEqual(None, validate)
```
Comment (emmadrigal) [inline tests/config/test_app_validator.py:506]: Good testing! it seems you are covering all if not most cases :+1:

File: tests/config/test_yaml.yaml [hunk 15]
```diff
@@ -0,0 +1,2 @@
+---
+test: 10
```
Comment (emmadrigal) [inline tests/config/test_yaml.yaml]: Add newline at the end of the file

File: rr/config/app_validator.py (hunk 10 above)
Comment (emmadrigal) [reply rr/config/app_validator.py]: Maybe even add the received type:
`"received a " stream_id.type() " expected a str"`
or
`stream_id "is a " stream_id.type() " expected a str"`

File: tests/config/test_yaml_invalid.yaml [hunk 16]
```diff
@@ -0,0 +1,5 @@
+---
+This is an invalid YAML:
+  - one
+  - two
+ - three (invalid indentation)
```
Comment (emmadrigal) [inline tests/config/test_yaml_invalid.yaml]: Add newline at the end of the file

**Review by emmadrigal — CHANGES_REQUESTED** [review]

File: rr/config/app_validator.py [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+        self.validate_triggers(cfg)
+
+    def validate_streams(self, cfg):
+        try:
+            streams = cfg['streams']
+        except KeyError as e:
+            raise AppValidatortError("streams object not found") from e
+
+        if (not isinstance(streams, list)):
+            raise AppValidatortError("Invalid streams format")
```
Comment (michaelgruner) [inline rr/config/app_validator.py]: Be consistent with the message formatting. Are you starting with upper case or lower case?

Comment (michaelgruner) [reply rr/config/app_validator.py]: Also, this message could be way more helpful. For example, consider something like: `found streams field, but it is not a list.`

Comment (michaelgruner) [reply rr/config/app_validator.py]: chain to the parent exception

File: rr/config/app_validator.py [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+        self.validate_actions(cfg)
+        self.validate_triggers(cfg)
+
+    def validate_streams(self, cfg):
+        try:
+            streams = cfg['streams']
+        except KeyError as e:
+            raise AppValidatortError("streams object not found") from e
+
+        if (not isinstance(streams, list)):
```
Comment (michaelgruner) [inline rr/config/app_validator.py]: There is no need for parenthesis in python.

File: rr/config/app_validator.py [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+            raise AppValidatortError("Invalid streams format")
+
+        for stream in streams:
+            try:
+                stream_id = stream['id']
```
Comment (michaelgruner) [inline rr/config/app_validator.py]: Note the code replication. Keep your code [dry](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself). You have the same pattern everywhere:

```python
try:
  value = dict[key]
except KeyError as e:
  raise AppValidatortError (key_err_msg) from e

if not isinstance(value, expected_type)):
  raise AppValidatortError(type_err_msg)
```
See how it could be factored out?

File: rr/config/app_validator.py [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+                    "triggers property not found in stream") from e
+
+            if (not isinstance(stream_id, str)):
+                raise AppValidatortError("Invalid id format")
+            if (not isinstance(stream_uri, str)):
+                raise AppValidatortError("Invalid uri format")
+            if (not isinstance(stream_triggers, list)):
+                raise AppValidatortError("Invalid triggers format")
+
+            for trigger in stream_triggers:
```
Comment (michaelgruner) [inline rr/config/app_validator.py]: How do we validate if the trigger actually exists?

File: rr/config/app_validator.py [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+            raise AppValidatortError("Invalid triggers format")
+
+        for trigger in triggers:
+            try:
+                trigger_name = trigger['name']
+            except KeyError as e:
+                raise AppValidatortError(
+                    "name property not found in triggers") from e
+            try:
+                trigger_action = trigger['action']
```
Comment (michaelgruner) [inline rr/config/app_validator.py]: Validate the action actually exists

File: rr/config/app_validator.py [hunk 22]
```diff
@@ -0,0 +1,175 @@
...
+            except KeyError as e:
+                raise AppValidatortError(
+                    "action property not found in triggers") from e
+            try:
+                trigger_filters = trigger['filters']
```
Comment (michaelgruner) [inline rr/config/app_validator.py]: Validate that the filter actually exists.

File: rr/config/app_validator.py [hunk 23]
```diff
@@ -0,0 +1,175 @@
...
+
+
+class AppValidatortError(RuntimeError):
+    pass
+
+
+class AppValidator():
+
+    def validate(self, cfg):
+        self.validate_streams(cfg)
```
Comment (michaelgruner) [inline rr/config/app_validator.py]: In other modules you prepend private methods with a `_` (which is a correct python standard). Please be consistent with the coding standard.

File: rr/config/config_loader.py [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+        self._validator = validator
+
+    def load(self, path):
+        cfg_obj = self._format.parse(path)
+        self._validator.validate(cfg_obj)
+
+        return cfg_obj
+
+
+class AppConfigLoader:
```
Comment (michaelgruner) [inline rr/config/config_loader.py]: Do not mix these. What if you need a new loader? the ConfigLoader is generic, while the AppConfigLoader is a specific facade.

**Review by michaelgruner — CHANGES_REQUESTED** [review]

File: rr/config/app_config_loader.py [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+#           Marisol Zeledon <marisol.zeledon@ridgerun.com>
+
+from rr.config.config_loader import ConfigLoader
+from rr.config.app_validator import AppValidator
+from rr.config.yaml_format import YamlFormat
+
+
+class AppConfigLoader:
+    """
+    Class that serves as the Apo Config facade
```
Comment (emmadrigal) [inline rr/config/app_config_loader.py]: Typo `App`

File: rr/config/app_validator.py [hunk 26]
```diff
@@ -0,0 +1,260 @@
...
+
+        trigger_filters = trigger['filters']
+        validate_lists(
+            trigger_filters,
+            str,
+            "Found filter element in trigger, but it is not a string")
+
+        for filt in trigger_filters:
+            validate_dependency(
+                filt, filters, "Trigger has a not defined filter")
```
Comment (michaelgruner) [inline rr/config/app_validator.py]: It would be helpful if you said which one is not defined. For example `Triggers attempts to use the \"%s\" filter, but it is not defined anywhere`.

File: rr/config/app_validator.py [hunk 27]
```diff
@@ -0,0 +1,260 @@
...
+        validate_dependency(
+            trigger['action'],
+            actions,
+            "Trigger has a not defined action")
```
Comment (michaelgruner) [inline rr/config/app_validator.py]: Same here.

File: rr/config/app_validator.py [hunk 28]
```diff
@@ -0,0 +1,260 @@
...
+            str,
+            "Found label element in filter, but it is not a string")
+
+        filters_list.append(filt['name'])
+
+    return filters_list
+
+
+def validate_actions(cfg):
+    """Validates the avtions field of the configuration object
```
Comment (emmadrigal) [inline rr/config/app_validator.py]: Typo `actions`

**Review by michaelgruner — APPROVED** [review]
Approved with a couple of comments. Please attend them before merging.

**Review by emmadrigal — APPROVED** [review]
